File: study1/week3/climbing-stairs.java
// https://leetcode.com/problems/climbing-stairs/
// Time complexity: O(n), n: input value
// Space complexity: O(1) because using only 3-size-array
class Solution {
    public int climbStairs(int n) {
        int[] temp = new int[]{1, 2};
        for (int i = 2; i < n; ++i) {
            temp[i % 2] = temp[0] + temp[1];
        }
        return temp[(n - 1) % 2];
    }
}

File: study1/solutions/week3/climbing_stairs.py
# https://leetcode.com/problems/climbing-stairs/description/
class Solution:
    def climbStairs(self, n: int) -> int:
        distinct_ways = [1, 2]    
        for i in range(2, n + 1):
            distinct_ways[i % 2] = distinct_ways[-1] + distinct_ways[-2]
        return distinct_ways[(n - 1) % 2]